mangaki/mangaki/utils/manami.py

```diff
@@ -0,0 +1,522 @@
+        self.from_synonym = {}
+        if manami_map is None:
+            manami_map = {k: k for k in range(MAX_MANAMI_ENTRIES)}
+        self.manami_map = manami_map
+
+        with open(self._path, encoding='utf-8') as f:
+            self._raw = json.load(f)
+            for i in range(len(self._raw['data'])):
+                self._raw['data'][i]['nb_episodes'] = str(self._raw['data'][i]['episodes'])
+                self._raw['data'][i]['subcategory'] = coarse_category[self._raw['data'][i]['type'].lower()]
+                self._raw['data'][i]['year'] = self._raw['data'][i]['animeSeason']['year'] if 'year' in self._raw['data'][i]['animeSeason'] else None
+
+        for local_id, datum in enumerate(self._raw['data']):
+            '''if local_id in EXCLUDED_MANAMI_IDS:
+                                                    continue'''
+            # Parse sources
+            for source in datum['sources']:
+                source, identifier = parse_manami_source(source)
+                if self._filter_sources is not None and source not in self._filter_sources:
+                    continue
+
+                self.references.setdefault((source, identifier), []).append(manami_map[local_id])
+                self._raw['data'][local_id].setdefault('references', []).append((source, identifier))
+
+            # Setup title reverse search
+            self.from_title.setdefault(sanitize(datum['title']), set()).add(manami_map[local_id])
+            for synonym in datum['synonyms']:
+                self.from_synonym.setdefault(sanitize(synonym), set()).add(manami_map[local_id])
+
+        self.df = pd.DataFrame(self._raw['data'])
+
+        self._check()
+
+    def _check(self):
+        for (source, identifier), manami_entry in self.references.items():
+            assert len(manami_entry) == 1, f"Multiple Manami entries with reference {source}/{identifier}"
+
+    def __getitem__(self, key):
+        return self._raw['data'][key]
+
+    def __len__(self):
+        return len(self._raw['data'])
+
+    def print_summary(self):
+        if self._filter_sources is None:
+            filters = ''
+        else:
+            filters = ' ({} only)'.format(', '.join(self._filter_sources))
+
+        print('Manami:')
+        print(f'    {len(self)} animes (with {len(self.from_title)} unique titles)')
+        print(f'    {len(self.references)} unique references{filters}')
+
+
+class MangakiDatabase:
+    def __init__(self, *, filter_sources=None):
+        self._raw = {work['pk']: work for work in Work.objects.filter(category__slug='anime').values(
+            'pk', 'title', 'nb_episodes', 'date', 'anime_type', 'nb_ratings')}
+        self._filter_sources = filter_sources
+        self.references = {}
+        self.from_title = {}
+        self.from_synonym = {}
+
+        # Extract all Mangaki references
+        redirects = dict(Work.all_objects.filter(redirect__isnull=False).values_list('pk', 'redirect_id'))
+
+        remember_anidb = {}
```

Comment:
C'est hélas compliqué à expliquer, je fais ça demain dans le code

Reply:
"""
    OK this is subtle. We usually do not allow duplicate refs,
    except for AniDB (e.g. Madoka Magica I, II, III). In this case,
    the first anime keeps the AniDB ref in this MangakiDatabase
    object, the other ones not (to avoid unnecessary merges).
    It works fine in 99% of cases.
    """

Reply:
cela dit c'est pas factorisé dans une fonction pls

Reply:
Fait. Pour Manami aussi.